pennylane_lightning/src/simulator/TransitionKernels.hpp

```diff
@@ -0,0 +1,112 @@
+#pragma once
+
+#include <algorithm>
+#include <complex>
+#include <cstdio>
+#include <random>
+#include <stack>
+#include <type_traits>
+#include <unordered_map>
+#include <vector>
+
+#include "StateVectorManagedCPU.hpp"
+#include "StateVectorRawCPU.hpp"
+
+namespace Pennylane {
+
+enum class TransitionKernelType { Local, NonRandom };
+
+template <typename fp_t> class TransitionKernel {
+  public:
+    virtual size_t init_state() = 0;
+    // outputs the next state and the qratio
+    virtual std::pair<size_t, fp_t> operator()(size_t) = 0;
+};
+
+template <typename fp_t>
+class LocalTransitionKernel : public TransitionKernel<fp_t> {
+  private:
+    std::random_device rd_;
+    std::mt19937 gen_;
+    std::uniform_int_distribution<size_t> distrib_num_qubits_;
+    std::uniform_int_distribution<size_t> distrib_binary_;
+    size_t num_qubits_;
+
+  public:
+    LocalTransitionKernel(size_t num_qubits) {
+        num_qubits_ = num_qubits;
+        gen_ = std::mt19937(rd_());
+        distrib_num_qubits_ =
+            std::uniform_int_distribution<size_t>(0, num_qubits - 1);
+        distrib_binary_ = std::uniform_int_distribution<size_t>(0, 1);
+    }
+
+    size_t init_state() { return 0; }
+
+    std::pair<size_t, fp_t> operator()(size_t s1) {
+        size_t qubit_site = distrib_num_qubits_(gen_);
+        size_t qubit_value = distrib_binary_(gen_);
+        size_t current_bit = (s1 >> qubit_site) & 1;
+
+        if (qubit_value == current_bit)
+            return std::pair<size_t, fp_t>(s1, 1);
+        else if (current_bit == 0) {
+            return std::pair<size_t, fp_t>(s1 + std::pow(2, qubit_site), 1);
+        } else {
+            return std::pair<size_t, fp_t>(s1 - std::pow(2, qubit_site), 1);
+        }
+    }
+};
+
+template <typename fp_t>
+class NonZeroRandomTransitionKernel : public TransitionKernel<fp_t> {
+    // : public TransitionKernel<fp_t> {
+  private:
+    std::random_device rd_;
+    std::mt19937 gen_;
+    std::uniform_int_distribution<size_t> distrib_;
+    size_t sv_length_;
+    std::vector<size_t> non_zeros_;
+
+  public:
+    size_t init_state() { return 0; }
```

Comment:
Why does this function exist again?

Reply:
This is a parent class method that needs to be defined in the child classes, even if it is "useless" per say. Let me know if there is a classier way to do this.